File: app/src/main/java/com/permissionnanny/operation/LocationOperation.java
package com.permissionnanny.operation;

import android.Manifest;
import android.content.Context;
import android.location.LocationManager;
import android.os.Bundle;
import com.permissionnanny.ProxyFunction;
import com.permissionnanny.R;
import com.permissionnanny.lib.request.RequestParams;
import com.permissionnanny.lib.request.simple.LocationRequest;

/**
 *
 */
public class LocationOperation {
    public static SimpleOperation[] operations = new SimpleOperation[]{
            new SimpleOperation(LocationRequest.ADD_GPS_STATUS_LISTENER,
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    R.string.dialogTitle_locationAddGpsStatusListener, 3, null),
            new SimpleOperation(LocationRequest.ADD_NMEA_LISTENER,
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    R.string.dialogTitle_locationAddNmeaListener, 5, null),
            new SimpleOperation(LocationRequest.ADD_PROXIMITY_ALERT,
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    R.string.dialogTitle_locationAddProximityAlert, 1, new ProxyFunction() {
                @Override
                public void execute(Context context, RequestParams request, Bundle response) {
                    LocationManager mgr = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
                    mgr.addProximityAlert(request.double0, request.double1, request.float0,
                            request.long0, request.pendingIntent0);
                }
            }),
            new SimpleOperation(LocationRequest.GET_LAST_KNOWN_LOCATION,
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    R.string.dialogTitle_locationGetLastKnownLocation, 1, new ProxyFunction() {
                @Override
                public void execute(Context context, RequestParams request, Bundle response) {
                    LocationManager mgr = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
                    response.putParcelable(request.opCode, mgr.getLastKnownLocation(request.string0));
                }
            }),
            new SimpleOperation(LocationRequest.REMOVE_GPS_STATUS_LISTENER,
                    null, R.string.dialogTitle_locationRemoveGpsStatusListener, 3, null),
            new SimpleOperation(LocationRequest.REMOVE_NMEA_LISTENER,
                    null, R.string.dialogTitle_locationRemoveNmeaListener, 5, null),
            new SimpleOperation(LocationRequest.REMOVE_PROXIMITY_ALERT,
                    null, R.string.dialogTitle_locationRemoveProximityAlert, 1, new ProxyFunction() {
                @Override
                public void execute(Context context, RequestParams request, Bundle response) {
                    LocationManager mgr = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
                    mgr.removeProximityAlert(request.pendingIntent0);
                }
            }),
            new SimpleOperation(LocationRequest.REMOVE_UPDATES,
                    null, R.string.dialogTitle_locationRemoveUpdates, 3, null),
            new SimpleOperation(LocationRequest.REMOVE_UPDATES1,
                    null, R.string.dialogTitle_locationRemoveUpdates1, 1, null),
            new SimpleOperation(LocationRequest.REQUEST_LOCATION_UPDATES,
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    R.string.dialogTitle_locationRequestLocationUpdates, 9, new ProxyFunction() {
                @Override
                public void execute(Context context, RequestParams request, Bundle response) {
                    LocationManager mgr = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
                    mgr.requestLocationUpdates(request.long0, request.float0, request.criteria0,
                            request.pendingIntent0);
                }
            }),
            new SimpleOperation(LocationRequest.REQUEST_LOCATION_UPDATES1,
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    R.string.dialogTitle_locationRequestLocationUpdates1, 9, null),
    };

    public static SimpleOperation getOperation(String opCode) {
        for (SimpleOperation operation : operations) {
            if (operation.mOpCode.equals(opCode)) {
                return operation;
            }
        }
        return null;
    }
}

File: app/src/main/kotlin/com/permissionnanny/simple/LocationOperation.kt
package com.permissionnanny.simple

import android.Manifest
import android.content.Context
import android.content.pm.PermissionInfo
import android.location.LocationManager
import com.permissionnanny.R
import com.permissionnanny.lib.request.simple.LocationRequest
import javax.inject.Inject

/**

 */
class LocationOperation
@Inject
constructor() {
    companion object {
        val operations = arrayOf(
                SimpleOperation(LocationRequest.ADD_GPS_STATUS_LISTENER,
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationAddGpsStatusListener, 3, null),
                SimpleOperation(LocationRequest.ADD_NMEA_LISTENER,
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationAddNmeaListener, 5, null),
                SimpleOperation(LocationRequest.ADD_PROXIMITY_ALERT,
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationAddProximityAlert, 1,
                        { context, request, response ->
                            val mgr = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                            mgr.addProximityAlert(request.double0, request.double1, request.float0,
                                    request.long0, request.pendingIntent0)
                        }),
                SimpleOperation(LocationRequest.GET_LAST_KNOWN_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationGetLastKnownLocation, 1,
                        { context, request, response ->
                            val mgr = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                            response.putParcelable(request.opCode, mgr.getLastKnownLocation(request.string0))
                        }),
                SimpleOperation(LocationRequest.REMOVE_PROXIMITY_ALERT,
                        "",
                        PermissionInfo.PROTECTION_NORMAL,
                        R.string.dialogTitle_locationRemoveProximityAlert, 1,
                        { context, request, response ->
                            val mgr = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                            mgr.removeProximityAlert(request.pendingIntent0)
                        }),
                SimpleOperation(LocationRequest.REMOVE_UPDATES,
                        "",
                        PermissionInfo.PROTECTION_NORMAL,
                        R.string.dialogTitle_locationRemoveUpdates, 3, null),
                SimpleOperation(LocationRequest.REQUEST_LOCATION_UPDATES,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9,
                        { context, request, response ->
                            val mgr = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                            mgr.requestLocationUpdates(request.long0, request.float0, request.criteria0,
                                    request.pendingIntent0)
                        }),
                SimpleOperation(LocationRequest.REQUEST_LOCATION_UPDATES1,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9, null),
                SimpleOperation(LocationRequest.REQUEST_LOCATION_UPDATES2,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9, null),
                SimpleOperation(LocationRequest.REQUEST_LOCATION_UPDATES3,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9, null),
                SimpleOperation(LocationRequest.REQUEST_SINGLE_UPDATE,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9, null),
                SimpleOperation(LocationRequest.REQUEST_SINGLE_UPDATE1,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9, null),
                SimpleOperation(LocationRequest.REQUEST_SINGLE_UPDATE2,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9,
                        { context, request, response ->
                            val mgr = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                            mgr.requestSingleUpdate(request.string0, request.pendingIntent0)
                        }),
                SimpleOperation(LocationRequest.REQUEST_SINGLE_UPDATE3,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        PermissionInfo.PROTECTION_DANGEROUS,
                        R.string.dialogTitle_locationRequestLocationUpdates, 9,
                        { context, request, response ->
                            val mgr = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                            mgr.requestSingleUpdate(request.criteria0, request.pendingIntent0)
                        })
        )

        fun getOperation(opCode: String): SimpleOperation? {
            for (operation in operations) {
                if (operation.opCode == opCode) {
                    return operation
                }
            }
            return null
        }
    }
}
